x/subscription/keeper/subscription.go

```diff
@@ -176,71 +123,100 @@ func (k Keeper) CreateSubscription(
 	}
 
-	// only one subscription per consumer
-	if _, found := k.GetSubscription(ctx, consumer); found {
-		details := map[string]string{"consumer": consumer}
-		return utils.LavaError(ctx, logger, "CreateSubscription", details, "consumer has existing subscription")
-	}
-
 	plan, found := k.plansKeeper.GetPlan(ctx, planIndex)
 	if !found {
 		details := map[string]string{
 			"plan":  planIndex,
-			"block": strconv.FormatInt(ctx.BlockHeight(), 10),
+			"block": strconv.FormatInt(int64(block), 10),
 		}
 		return utils.LavaError(ctx, logger, "CreateSubscription", details, "invalid plan")
 	}
 
-	err = k.projectsKeeper.CreateDefaultProject(ctx, consumer)
-	if err != nil {
-		details := map[string]string{
-			"err": err.Error(),
+	sub, found := k.GetSubscription(ctx, consumer)
+
+	// Subscription creation:
+	//   When: if not already exists for consumer address)
+	//   What: find plan, create default project, set duration, calculate price,
+	//         charge fees, save subscription.
+	//
+	// Subscription renewal:
+	//   When: if already exists and existing plan is the same as current plans
+	//         ("same" means same index and same block of creation)
+	//   What: find plan, update duration (total and remaining), calculate price,
+	//         charge fees, save subscription.
+	//
+	// Subscription upgrade: (TBD)
+	//
+	// Subscription downgrade: (TBD)
+
+	if !found {
+		// creeate new subscription with this plan
+		sub = types.Subscription{
+			Creator:   creator,
+			Consumer:  consumer,
+			Block:     block,
+			PlanIndex: planIndex,
+			PlanBlock: plan.Block,
+		}
+
+		sub.MonthCuTotal = plan.GetComputeUnits()
+		sub.MonthCuLeft = plan.GetComputeUnits()
+
+		// new subscription needs a default project
+		if err = k.projectsKeeper.CreateDefaultProject(ctx, consumer); err != nil {
+			details := map[string]string{
+				"err": err.Error(),
+			}
+			return utils.LavaError(ctx, logger, "CreateSubscription", details, "failed to create default project")
+		}
+	} else {
+		// allow renewal with the same plan ("same" means both plan index,block match);
+		// otherwise, only one subscription per consumer
+		if !(plan.Index == sub.PlanIndex && plan.Block == sub.PlanBlock) {
+			details := map[string]string{"consumer": consumer}
+			return utils.LavaError(ctx, logger, "CreateSubscription", details, "consumer has existing subscription")
+		}
+
+		// For now, allow renewal only by the same creator.
+		// TODO: after adding fixation, we can allow different creators
+		if creator != sub.Creator {
+			details := map[string]string{"creator": consumer}
+			return utils.LavaError(ctx, logger, "CreateSubscription", details, "existing subscription has different creator")
+		}
+
+		// The total duration may not exceed MAX_SUBSCRIPTION_DURATION, but allow an
+		// extra month to account for renwewals before the end of current subscription
+		if sub.DurationLeft+duration > types.MAX_SUBSCRIPTION_DURATION+1 {
```

Comment:
if sub.DurationLeft+duration >= types.MAX_SUBSCRIPTION_DURATION

Reply:
i get it, im just saying that this is confusing, why not remove the +1 and use >= same result, more readable

Reply:
They aren't equivalent... 
(DurationLeft=1 with duration=11 (or 12) are both valid, but 1+11 >= 12 (or 13) both evaluate to true)
Besides, the explicit "+1" makes the logic visible: MAX plus ONE